src/Decompiler/Analysis/CallRewriter.cs

```diff
@@ -483,16 +479,18 @@ public void RewriteReturns(SsaState ssa)
             SsaState ssa,
             Block block,
             Identifier idRet,
-            Expression e)
+            CallBinding idStg)
         {
+            var e = idStg?.Expression ?? Constant.Invalid;
             for (int i = block.Statements.Count-1; i >=0; --i)
             {
                 var stm = block.Statements[i];
                 if (stm.Instruction is ReturnInstruction ret)
                 {
                     if (idRet.DataType.BitSize < e.DataType.BitSize)
                     {
-                        e = new Cast(idRet.DataType, e);
```

Comment:
Nice catch. This was a bug before, no attention was being paid to the _offset_ of the smaller datatype.